src/xsdba/utils.py

```diff
@@ -0,0 +1,1238 @@
+
+    A cluster is defined as a sequence of values larger than u2 with at least one value larger than u1.
+
+    Parameters
+    ----------
+    data : 1D ndarray
+      Values to get clusters from.
+    u1 : float
+      Extreme value threshold, at least one value in the cluster must exceed this.
+    u2 : float
+      Cluster threshold, values above this can be part of a cluster.
+
+    Returns
+    -------
+    (np.array, np.array, np.array, np.array)
+
+    References
+    ----------
+    `getcluster` of Extremes.jl (:cite:cts:`jalbert_extreme_2022`).
+    """
+    # Boolean array, True where data is over u2
+    # We pad with values under u2, so that clusters never start or end at boundaries.
+    exce = np.concatenate(([u2 - 1], data, [u2 - 1])) > u2
+
+    # 1 just before the start of the cluster
+    # -1 on the last element of the cluster
+    bounds = np.diff(exce.astype(np.int32))
+    # We add 1 to get the first element and sub 1 to get the same index as in data
+    starts = np.where(bounds == 1)[0]
+    # We sub 1 to get the same index as in data and add 1 to get the element after (for python slicing)
+    ends = np.where(bounds == -1)[0]
+
+    cl_maxpos = []
+    cl_maxval = []
+    cl_start = []
+    cl_end = []
+    for start, end in zip(starts, ends, strict=False):
+        cluster_max = data[start:end].max()
+        if cluster_max > u1:
+            cl_maxval.append(cluster_max)
+            cl_maxpos.append(start + np.argmax(data[start:end]))
+            cl_start.append(start)
+            cl_end.append(end - 1)
+
+    return (
+        np.array(cl_start),
+        np.array(cl_end),
+        np.array(cl_maxpos),
+        np.array(cl_maxval),
+    )
+
+
+def get_clusters(data: xr.DataArray, u1, u2, dim: str = "time") -> xr.Dataset:
+    """Get cluster count, maximum and position along a given dim.
+
+    See `get_clusters_1d`. Used by `adjustment.ExtremeValues`.
+
+    Parameters
+    ----------
+    data: 1D ndarray
+      Values to get clusters from.
+    u1 : float
+      Extreme value threshold, at least one value in the cluster must exceed this.
+    u2 : float
+      Cluster threshold, values above this can be part of a cluster.
+    dim : str
+      Dimension name.
+
+    Returns
+    -------
+    xr.Dataset
+      With variables,
+        - `nclusters` : Number of clusters for each point (with `dim` reduced), int
+        - `start` : First index in the cluster (`dim` reduced, new `cluster`), int
+        - `end` : Last index in the cluster, inclusive (`dim` reduced, new `cluster`), int
+        - `maxpos` : Index of the maximal value within the cluster (`dim` reduced, new `cluster`), int
+        - `maximum` : Maximal value within the cluster (`dim` reduced, new `cluster`), same dtype as data.
+
+      For `start`, `end` and `maxpos`, -1 means nan and should always correspond to a `nan` in `maximum`.
+      The length along `cluster` is half the size of "dim", the maximal theoretical number of clusters.
```

Comment:
```suggestion
    data: 1D ndarray
        Values to get clusters from.
    u1 : float
        Extreme value threshold, at least one value in the cluster must exceed this.
    u2 : float
        Cluster threshold, values above this can be part of a cluster.
    dim : str
        Dimension name.

    Returns
    -------
    xr.Dataset
        With variables,
            - `nclusters` : Number of clusters for each point (with `dim` reduced), int
            - `start` : First index in the cluster (`dim` reduced, new `cluster`), int
            - `end` : Last index in the cluster, inclusive (`dim` reduced, new `cluster`), int
            - `maxpos` : Index of the maximal value within the cluster (`dim` reduced, new `cluster`), int
            - `maximum` : Maximal value within the cluster (`dim` reduced, new `cluster`), same dtype as data.

        For `start`, `end` and `maxpos`, -1 means nan and should always correspond to a `nan` in `maximum`.
        The length along `cluster` is half the size of "dim", the maximal theoretical number of clusters.
```